src/main/java/gregtech/common/covers/ender/CoverAbstractEnderLink.java

```diff
@@ -0,0 +1,495 @@
+import org.jetbrains.annotations.Nullable;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.UUID;
+import java.util.regex.Pattern;
+
+@SuppressWarnings("SameParameterValue")
+public abstract class CoverAbstractEnderLink<T extends VirtualEntry> extends CoverBase
+                                            implements CoverWithUI, ITickable, IControllable {
+
+    protected static final Pattern COLOR_INPUT_PATTERN = Pattern.compile("[0-9a-fA-F]*");
+    public static final int UPDATE_PRIVATE = GregtechDataCodes.assignId();
+
+    protected T activeEntry = null;
+    protected String color = VirtualEntry.DEFAULT_COLOR;
+    protected UUID playerUUID = null;
+    private boolean isPrivate = false;
+    private boolean workingEnabled = true;
+    private boolean ioEnabled = false;
+
+    public CoverAbstractEnderLink(@NotNull CoverDefinition definition, @NotNull CoverableView coverableView,
+                                  @NotNull EnumFacing attachedSide) {
+        super(definition, coverableView, attachedSide);
+        updateLink();
+    }
+
+    protected void updateLink() {
+        this.activeEntry = VirtualEnderRegistry.getOrCreateEntry(getOwner(), getType(), createName());
+        this.activeEntry.setColor(this.color);
+        markDirty();
+    }
+
+    protected abstract EntryTypes<T> getType();
+
+    public String getColorStr() {
+        return this.color;
+    }
+
+    protected final String createName() {
+        return identifier() + this.color;
+    }
+
+    protected abstract String identifier();
+
+    protected final UUID getOwner() {
+        return isPrivate ? playerUUID : null;
+    }
+
+    @Override
+    public void readCustomData(int discriminator, @NotNull PacketBuffer buf) {
+        super.readCustomData(discriminator, buf);
+        if (discriminator == UPDATE_PRIVATE) {
+            setPrivate(buf.readBoolean());
+            updateLink();
+        }
+    }
+
+    @Override
+    public @NotNull EnumActionResult onScrewdriverClick(@NotNull EntityPlayer playerIn, @NotNull EnumHand hand,
+                                                        @NotNull CuboidRayTraceResult hitResult) {
+        if (!getWorld().isRemote) {
+            openUI((EntityPlayerMP) playerIn);
+        }
+        return EnumActionResult.SUCCESS;
+    }
+
+    @Override
+    public void onAttachment(@NotNull CoverableView coverableView, @NotNull EnumFacing side,
+                             @Nullable EntityPlayer player, @NotNull ItemStack itemStack) {
+        super.onAttachment(coverableView, side, player, itemStack);
+        if (player != null) {
+            this.playerUUID = player.getUniqueID();
+        }
+    }
+
+    public void updateColor(String str) {
+        if (str.length() == 8) {
+            this.color = str.toUpperCase();
+            updateLink();
+        }
```

Comment:
Should we throw when false, or use an assert instead of the if statement altogether?

Reply:
previous logic would just not accept the malformed string and stay on the same color. i don't think throwing would be ideal for user input